encoding/cbf/cbf_codec/encode_value.go

```diff
@@ -0,0 +1,394 @@
+	case cadence.Struct:
+		err = e.EncodeValueIdentifier(EncodedValueStruct)
+		if err != nil {
+			return
+		}
+		return e.EncodeStruct(v)
+	case cadence.Resource:
+		err = e.EncodeValueIdentifier(EncodedValueResource)
+		if err != nil {
+			return
+		}
+		return e.EncodeResource(v)
+	case cadence.Event:
+		err = e.EncodeValueIdentifier(EncodedValueEvent)
+		if err != nil {
+			return
+		}
+		return e.EncodeEvent(v)
+	case cadence.Contract:
+		err = e.EncodeValueIdentifier(EncodedValueContract)
+		if err != nil {
+			return
+		}
+		return e.EncodeContract(v)
+	case cadence.Link:
+		err = e.EncodeValueIdentifier(EncodedValueLink)
+		if err != nil {
+			return
+		}
+		return e.EncodeLink(v)
+	case cadence.Path:
+		err = e.EncodeValueIdentifier(EncodedValuePath)
+		if err != nil {
+			return
+		}
+		return e.EncodePath(v)
+	case cadence.Capability:
+		err = e.EncodeValueIdentifier(EncodedValueCapability)
+		if err != nil {
+			return
+		}
+		return e.EncodeCapability(v)
+	case cadence.Enum:
+		err = e.EncodeValueIdentifier(EncodedValueEnum)
+		if err != nil {
+			return
+		}
+		return e.EncodeEnum(v)
+	}
+
+	return common_codec.CodecError(fmt.Sprintf("unexpected value: %s (type=%s)", value, value.Type()))
+}
+
+func (e *Encoder) EncodeValueIdentifier(id EncodedValue) (err error) {
+	return e.writeByte(byte(id))
+}
+
+func (e *Encoder) EncodeOptional(value cadence.Optional) (err error) {
+	isNil := value.Value == nil
+	err = common_codec.EncodeBool(&e.w, isNil)
+	if isNil || err != nil {
+		return
+	}
+
+	return e.EncodeValue(value.Value)
+}
+
+func (e *Encoder) EncodeArray(value cadence.Array) (err error) {
+	switch v := value.ArrayType.(type) {
+	case cadence.VariableSizedArrayType, nil: // unknown type still needs length
+		err = common_codec.EncodeLength(&e.w, len(value.Values))
+		if err != nil {
+			return
+		}
+	case cadence.ConstantSizedArrayType:
+		if len(value.Values) != int(v.Size) {
+			return common_codec.CodecError(fmt.Sprintf("constant size array size=%d but has %d elements", v.Size, len(value.Values)))
+		}
+	}
+
+	for _, element := range value.Values {
+		err = e.EncodeValue(element)
+		if err != nil {
+			return err
+		}
+	}
+
+	return
+}
+
+func (e *Encoder) EncodeDictionary(value cadence.Dictionary) (err error) {
+	err = e.EncodeDictionaryType(value.DictionaryType)
+	if err != nil {
+		return
+	}
+	err = common_codec.EncodeLength(&e.w, len(value.Pairs))
+	if err != nil {
+		return
+	}
+	// TODO if dictionary type is concrete for key or value, don't encode type info for them
+	for _, kv := range value.Pairs {
+		err = e.EncodeValue(kv.Key)
+		if err != nil {
+			return
+		}
+		err = e.EncodeValue(kv.Value)
+		if err != nil {
+			return
+		}
+	}
+	return
+}
+
+func (e *Encoder) EncodeStruct(value cadence.Struct) (err error) {
+	err = e.EncodeStructType(value.StructType)
+	if err != nil {
+		return
+	}
+
+	return EncodeArray(e, value.Fields, func(field cadence.Value) (err error) {
+		return e.EncodeValue(field)
+	})
```

Comment:
```suggestion
	return EncodeArray(e, value.Fields, e.EncodeValue)
```

eta reduce: (𝜆𝑥.𝑀𝑥)𝑁⇝𝑀𝑁